wagtail/wagtailadmin/views/mixins.py

```diff
@@ -0,0 +1,45 @@
+from __future__ import absolute_import, unicode_literals
+
+from django.utils.translation import ugettext as _
+
+from wagtail.wagtailadmin.forms import SearchForm
+from wagtail.wagtailsearch.backends import get_search_backend
+from wagtail.wagtailsearch.index import class_is_indexed
+
+
+class SearchableListMixin(object):
+    search_box_placeholder = _("Search")
+    search_fields = None
+
+    def get_search_form(self):
+        if self.request.GET.get('q'):
+            return SearchForm(self.request.GET, placeholder=self.search_box_placeholder)
+        else:
+            return SearchForm(placeholder=self.search_box_placeholder)
+
+    def get_queryset(self):
+        queryset = super(SearchableListMixin, self).get_queryset()
+        search_form = self.get_search_form()
+
+        if search_form.is_valid():
+            q = search_form.cleaned_data['q']
+
+            if class_is_indexed(queryset.model):
+                search_backend = get_search_backend()
+                queryset = search_backend.search(q, queryset, fields=self.search_fields)
+            else:
+                filters = {
+                    field + '__icontains': q
+                    for field in self.search_fields or []
+                }
+
+                queryset = queryset.filter(**filters)
+
+        return queryset
```

Comment:
Defining `queryset` is useless here, we can use `return` directly and avoid the `else:`.

Reply:
I'm not sure what you mean, this method should always return a queryset.

Reply:
Yes, it returns a queryset, but it does not have to define a `queryset` variable, querysets can be returned directly.